File: duna/Boj9012.py
import sys
input=sys.stdin.readline

n=int(input())
for _ in range(n):
  stack=[]
  arr=input().strip()
  if arr[0]==')' or arr[-1]=='(':
    print("NO")
  else:
    stack.append(arr[0])
  
    for i in range(1,len(arr)):
      #print("stack=",stack)
      if len(stack)==0 and arr[i]==')':
        stack.append(-1)
        break;
      elif len(stack)>0 and stack[-1] != arr[i]:
        stack.pop()
      else:
        stack.append(arr[i])

    print("YES" if len(stack)==0 else "NO")

File: suyeon/자료구조1/Boj9012.java
package suyeon.자료구조1;

import java.io.*;
import java.util.Stack;
//백준 9012 괄호
public class Boj9012 {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();

        //입력 받을 테스트 케이스의 수
        int testCaseCount = Integer.parseInt(br.readLine());

        for (int i = 0; i < testCaseCount; i++) {
            sb.append(solution(br.readLine())).append("\n");
        }
        System.out.println(sb);
    }

    static String solution(String inputPS){

        Stack<Character> stack = new Stack<>();

        for (int i = 0; i < inputPS.length(); i++) {

            //입력받은 문자열을 char 형으로 쪼개준다.
            char pv = inputPS.charAt(i);

            //여는 괄호 ( 를 스택에 담는다.
            if(pv == '(') stack.push(pv);
            //닫는 괄호 ) 를 담으려고 했으나 비어있으면 NO
            else if(stack.empty()) return "NO";
            else stack.pop();

        }

        if(stack.empty()) return "YES";
        else return "NO";

     }

}
